// ariakit/esm/combobox/combobox-item.js
import { useCallback } from 'react';
import { isTextField, getPopupItemRole } from 'ariakit-utils/dom';
import { isDownloading, isOpeningInNewTab } from 'ariakit-utils/events';
import { hasFocus } from 'ariakit-utils/focus';
import { useBooleanEvent, useEvent, useWrapElement } from 'ariakit-utils/hooks';
import { queueMicrotask } from 'ariakit-utils/misc';
import { useStore, createMemoComponent } from 'ariakit-utils/store';
import { createHook, createElement } from 'ariakit-utils/system';
import { useCompositeHover } from '../composite/composite-hover.js';
import { useCompositeItem } from '../composite/composite-item.js';
import { C as ComboboxContext, a as ComboboxItemValueContext } from '../__utils-8b39082b.js';
import { jsx } from 'react/jsx-runtime';

/**
 * A component hook that returns props that can be passed to `Role` or any other
 * Ariakit component to render a combobox item.
 * @see https://ariakit.org/components/combobox
 * @example
 * ```jsx
 * const state = useComboboxState();
 * const props = useComboboxItem({ state, value: "value" });
 * <Role {...props} />
 * ```
 */
const useComboboxItem = createHook(_ref => {
  var _state6, _state7;

  let {
    state,
    value,
    hideOnClick = value != null,
    setValueOnClick = true,
    shouldRegisterItem = true,
    focusOnHover = false,
    moveOnKeyPress = true,
    getItem: getItemProp,
    ...props
  } = _ref;
  state = useStore(state || ComboboxContext, ["setValue", "move", "hide", "baseRef", "contentElement", "mounted"]);
  const getItem = useCallback(item => {
    const nextItem = { ...item,
      value
    };

    if (getItemProp) {
      return getItemProp(nextItem);
    }

    return nextItem;
  }, [value, getItemProp]);
  const onClickProp = props.onClick;
  const setValueOnClickProp = useBooleanEvent(setValueOnClick);
  const hideOnClickProp = useBooleanEvent(hideOnClick);
  const onClick = useEvent(event => {
    onClickProp == null ? void 0 : onClickProp(event);
    if (event.defaultPrevented) return;
    if (isDownloading(event)) return;
    if (isOpeningInNewTab(event)) return;

    if (value != null && setValueOnClickProp(event)) {
      var _state;

      (_state = state) == null ? void 0 : _state.setValue(value);
    }

    if (hideOnClickProp(event)) {
      var _state2, _state3;

      // When ComboboxList is used instead of ComboboxPopover, state.hide()
      // does nothing. The focus will not be moved over to the combobox
      // input automatically. So we need to move manually here.
      (_state2 = state) == null ? void 0 : _state2.move(null);
      (_state3 = state) == null ? void 0 : _state3.hide();
    }
  });
  const onKeyDownProp = props.onKeyDown;
  const onKeyDown = useEvent(event => {
    var _state4;

    onKeyDownProp == null ? void 0 : onKeyDownProp(event);
    if (event.defaultPrevented) return;
    const baseElement = (_state4 = state) == null ? void 0 : _state4.baseRef.current;
    if (!baseElement) return;
    if (hasFocus(baseElement)) return; // When the combobox is not working with virtual focus, the items will
    // receive DOM focus. Therefore, pressing printable keys will not fill
    // the text field. So we need to programmatically focus on the text
    // field when the user presses printable keys.

    const printable = event.key.length === 1;

    if (printable || event.key === "Backspace" || event.key === "Delete") {
      queueMicrotask(() => baseElement.focus());

      if (isTextField(baseElement)) {
        var _state5;

        // If the combobox element is a text field, we should update the
        // state value with the current's element value. This is necessary
        // because the value may temporarily change based on the currently
        // selected item, but it'll be reset to the original value when the
        // combobox input is focused.
        (_state5 = state) == null ? void 0 : _state5.setValue(baseElement.value);
      }
    }
  });
  props = useWrapElement(props, element => /*#__PURE__*/jsx(ComboboxItemValueContext.Provider, {
    value: value,
    children: element
  }), [value]);
  props = {
    role: getPopupItemRole((_state6 = state) == null ? void 0 : _state6.contentElement),
    children: value,
    ...props,
    onClick,
    onKeyDown
  };
  const moveOnKeyPressProp = useBooleanEvent(moveOnKeyPress);
  props = useCompositeItem({
    state,
    ...props,
    getItem,
    // We only register the item on the state when the popover is open so we
    // don't try to move focus to hidden items when pressing arrow keys.
    shouldRegisterItem: ((_state7 = state) == null ? void 0 : _state7.mounted) && shouldRegisterItem,
    // Dispatch a custom event on the combobox input when moving to an item
    // with the keyboard so the Combobox component can enable inline
    // autocompletion.
    moveOnKeyPress: event => {
      var _state8, _state8$baseRef$curre;

      if (!moveOnKeyPressProp(event)) return false;
      const moveEvent = new Event("combobox-item-move");
      (_state8 = state) == null ? void 0 : (_state8$baseRef$curre = _state8.baseRef.current) == null ? void 0 : _state8$baseRef$curre.dispatchEvent(moveEvent);
      return true;
    }
  });
  props = useCompositeHover({
    state,
    focusOnHover,
    ...props
  });
  return props;
});
/**
 * A component that renders a combobox item inside a combobox list or popover.
 * The `role` prop will be automatically set based on the `ComboboxList` or
 * `ComboboxPopover` own `role` prop. For example, if the `ComboboxPopover`
 * component's `role` prop is set to `listbox` (default), the `ComboboxItem`
 * `role` will be set to `option`. By default, the `value` prop will be rendered
 * as the children, but this can be overriden.
 * @see https://ariakit.org/components/combobox
 * @example
 * ```jsx
 * const combobox = useComboboxState();
 * <Combobox state={combobox} />
 * <ComboboxPopover state={combobox}>
 *   <ComboboxItem value="Item 1" />
 *   <ComboboxItem value="Item 2" />
 *   <ComboboxItem value="Item 3" />
 * </ComboboxPopover>
 * ```
 */

const ComboboxItem = createMemoComponent(props => {
  const htmlProps = useComboboxItem(props);
  return createElement("div", htmlProps);
});

if (process.env.NODE_ENV !== "production") {
  ComboboxItem.displayName = "ComboboxItem";
}

export { ComboboxItem, useComboboxItem };
