// ariakit/cjs/popover/popover.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var react = require('react');
var hooks = require('ariakit-utils/hooks');
var system = require('ariakit-utils/system');
var dialog_dialog = require('../dialog/dialog.js');
var __utils = require('../__utils-bb7fa552.js');
var jsxRuntime = require('react/jsx-runtime');

/**
 * A component hook that returns props that can be passed to `Role` or any other
 * Ariakit component to render a popover element.
 * @see https://ariakit.org/components/popover
 * @example
 * ```jsx
 * const state = usePopoverState();
 * const props = usePopover({ state });
 * <Role {...props}>Popover</Role>
 * ```
 */
const usePopover = system.createHook(_ref => {
  let {
    state,
    modal = false,
    portal = !!modal,
    preserveTabOrder = true,
    autoFocusOnShow = true,
    wrapperProps,
    ...props
  } = _ref;
  const popoverRef = state.popoverRef; // Makes sure the wrapper element that's passed to popper has the same
  // z-index as the popover element so users only need to set the z-index
  // once.

  hooks.useSafeLayoutEffect(() => {
    const wrapper = popoverRef.current;
    const popover = state.contentElement;
    if (!wrapper) return;
    if (!popover) return;
    wrapper.style.zIndex = getComputedStyle(popover).zIndex;
  }, [popoverRef, state.contentElement]); // We have to wait for the popover to be positioned before we can move
  // focus, otherwise there may be scroll jumps. See popover-standalone
  // example test-browser file.

  const [canAutoFocusOnShow, setCanAutoFocusOnShow] = react.useState(false);
  const {
    portalRef,
    domReady
  } = hooks.usePortalRef(portal, props.portalRef);
  hooks.useSafeLayoutEffect(() => {
    var _state$contentElement;

    if (!domReady) return;
    if (!state.mounted) return;
    if (!((_state$contentElement = state.contentElement) != null && _state$contentElement.isConnected)) return;
    const raf = requestAnimationFrame(() => {
      setCanAutoFocusOnShow(true);
    });
    return () => {
      cancelAnimationFrame(raf);
    };
  }, [domReady, state.mounted, state.contentElement]); // Wrap our element in a div that will be used to position the popover.
  // This way the user doesn't need to override the popper's position to
  // create animations.

  props = hooks.useWrapElement(props, element => /*#__PURE__*/jsxRuntime.jsx("div", {
    role: "presentation",
    ...wrapperProps,
    style: {
      position: state.fixed ? "fixed" : "absolute",
      top: 0,
      left: 0,
      ...(wrapperProps == null ? void 0 : wrapperProps.style)
    },
    ref: popoverRef,
    children: element
  }), [state.fixed, popoverRef, wrapperProps]);
  props = hooks.useWrapElement(props, element => /*#__PURE__*/jsxRuntime.jsx(__utils.PopoverContext.Provider, {
    value: state,
    children: element
  }), [state]);
  props = { ...props,
    style: {
      position: "relative",
      ...props.style
    }
  };
  props = dialog_dialog.useDialog({
    state,
    modal,
    preserveTabOrder,
    portal,
    autoFocusOnShow: canAutoFocusOnShow && autoFocusOnShow,
    ...props,
    portalRef
  });
  return props;
});
/**
 * A component that renders a popover element.
 * @see https://ariakit.org/components/popover
 * @example
 * ```jsx
 * const popover = usePopoverState();
 * <PopoverDisclosure state={popover}>Disclosure</PopoverDisclosure>
 * <Popover state={popover}>Popover</Popover>
 * ```
 */

const Popover = system.createComponent(props => {
  const htmlProps = usePopover(props);
  return system.createElement("div", htmlProps);
});

if (process.env.NODE_ENV !== "production") {
  Popover.displayName = "Popover";
}

exports.Popover = Popover;
exports.usePopover = usePopover;
