// ariakit-utils/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var array = require('./array.js');
var dom = require('./dom.js');
var events = require('./events.js');
var focus = require('./focus.js');
var hooks = require('./hooks.js');
var misc = require('./misc.js');
var platform = require('./platform.js');
var store = require('./store.js');
var system = require('./system.js');



exports.addItemToArray = array.addItemToArray;
exports.flatten2DArray = array.flatten2DArray;
exports.reverseArray = array.reverseArray;
exports.toArray = array.toArray;
exports.canUseDOM = dom.canUseDOM;
exports.closest = dom.closest;
exports.contains = dom.contains;
exports.getActiveElement = dom.getActiveElement;
exports.getDocument = dom.getDocument;
exports.getPopupItemRole = dom.getPopupItemRole;
exports.getPopupRole = dom.getPopupRole;
exports.getScrollingElement = dom.getScrollingElement;
exports.getTextboxSelection = dom.getTextboxSelection;
exports.getWindow = dom.getWindow;
exports.isButton = dom.isButton;
exports.isFrame = dom.isFrame;
exports.isPartiallyHidden = dom.isPartiallyHidden;
exports.isTextField = dom.isTextField;
exports.isVisible = dom.isVisible;
exports.matches = dom.matches;
exports.scrollIntoViewIfNeeded = dom.scrollIntoViewIfNeeded;
exports.addGlobalEventListener = events.addGlobalEventListener;
exports.fireBlurEvent = events.fireBlurEvent;
exports.fireClickEvent = events.fireClickEvent;
exports.fireEvent = events.fireEvent;
exports.fireFocusEvent = events.fireFocusEvent;
exports.fireKeyboardEvent = events.fireKeyboardEvent;
exports.isDownloading = events.isDownloading;
exports.isFocusEventOutside = events.isFocusEventOutside;
exports.isOpeningInNewTab = events.isOpeningInNewTab;
exports.isPortalEvent = events.isPortalEvent;
exports.isSelfTarget = events.isSelfTarget;
exports.queueBeforeEvent = events.queueBeforeEvent;
exports.disableFocus = focus.disableFocus;
exports.disableFocusIn = focus.disableFocusIn;
exports.focusIfNeeded = focus.focusIfNeeded;
exports.focusIntoView = focus.focusIntoView;
exports.getAllFocusable = focus.getAllFocusable;
exports.getAllFocusableIn = focus.getAllFocusableIn;
exports.getAllTabbable = focus.getAllTabbable;
exports.getAllTabbableIn = focus.getAllTabbableIn;
exports.getClosestFocusable = focus.getClosestFocusable;
exports.getFirstFocusable = focus.getFirstFocusable;
exports.getFirstFocusableIn = focus.getFirstFocusableIn;
exports.getFirstTabbable = focus.getFirstTabbable;
exports.getFirstTabbableIn = focus.getFirstTabbableIn;
exports.getLastTabbable = focus.getLastTabbable;
exports.getLastTabbableIn = focus.getLastTabbableIn;
exports.getNextTabbable = focus.getNextTabbable;
exports.getNextTabbableIn = focus.getNextTabbableIn;
exports.getPreviousTabbable = focus.getPreviousTabbable;
exports.getPreviousTabbableIn = focus.getPreviousTabbableIn;
exports.hasFocus = focus.hasFocus;
exports.hasFocusWithin = focus.hasFocusWithin;
exports.isFocusable = focus.isFocusable;
exports.isTabbable = focus.isTabbable;
exports.restoreFocusIn = focus.restoreFocusIn;
exports.useBooleanEvent = hooks.useBooleanEvent;
exports.useControlledState = hooks.useControlledState;
exports.useDeferredValue = hooks.useDeferredValue;
exports.useEvent = hooks.useEvent;
exports.useForceUpdate = hooks.useForceUpdate;
exports.useForkRef = hooks.useForkRef;
exports.useId = hooks.useId;
exports.useInitialValue = hooks.useInitialValue;
exports.useIsMouseMoving = hooks.useIsMouseMoving;
exports.useLazyValue = hooks.useLazyValue;
exports.useLiveRef = hooks.useLiveRef;
exports.usePortalRef = hooks.usePortalRef;
exports.usePreviousValue = hooks.usePreviousValue;
exports.useRefId = hooks.useRefId;
exports.useSafeLayoutEffect = hooks.useSafeLayoutEffect;
exports.useTagName = hooks.useTagName;
exports.useUpdateEffect = hooks.useUpdateEffect;
exports.useUpdateLayoutEffect = hooks.useUpdateLayoutEffect;
exports.useWrapElement = hooks.useWrapElement;
exports.applyState = misc.applyState;
exports.chain = misc.chain;
exports.cx = misc.cx;
exports.hasOwnProperty = misc.hasOwnProperty;
exports.isEmpty = misc.isEmpty;
exports.isInteger = misc.isInteger;
exports.isObject = misc.isObject;
exports.isPromise = misc.isPromise;
exports.noop = misc.noop;
exports.normalizeString = misc.normalizeString;
exports.queueMicrotask = misc.queueMicrotask;
exports.setRef = misc.setRef;
exports.shallowEqual = misc.shallowEqual;
exports.isApple = platform.isApple;
exports.isFirefox = platform.isFirefox;
exports.isMac = platform.isMac;
exports.isSafari = platform.isSafari;
exports.isTouchDevice = platform.isTouchDevice;
exports.createMemoComponent = store.createMemoComponent;
exports.createStoreContext = store.createStoreContext;
exports.useStore = store.useStore;
exports.useStoreProvider = store.useStoreProvider;
exports.useStorePublisher = store.useStorePublisher;
exports.createComponent = system.createComponent;
exports.createElement = system.createElement;
exports.createHook = system.createHook;
