// gatsby-plugin-next-seo/lib/__mocks__/ssr-context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFilledContext = exports.context = void 0;
var helmet = {
  htmlAttributes: {
    toComponent: function toComponent() {
      return "html-attributes-component";
    }
  },
  bodyAttributes: {
    toComponent: function toComponent() {
      return "body-attributes-component";
    }
  },
  title: {
    toComponent: function toComponent() {
      return "title-component";
    }
  },
  link: {
    toComponent: function toComponent() {
      return "link-component";
    }
  },
  meta: {
    toComponent: function toComponent() {
      return "meta-component";
    }
  },
  noscript: {
    toComponent: function toComponent() {
      return "noscript-component";
    }
  },
  script: {
    toComponent: function toComponent() {
      return "script-component";
    }
  },
  style: {
    toComponent: function toComponent() {
      return "style-component";
    }
  },
  base: {
    toComponent: function toComponent() {
      return "base-component";
    }
  }
};
var context = {};
exports.context = context;

var getFilledContext = function getFilledContext() {
  return {
    helmet: helmet
  };
};

exports.getFilledContext = getFilledContext;
